pkg/ovsdb/ovsdb.go

```diff
@@ -0,0 +1,329 @@
+// Modifications copyright (C) 2017 Che Wei, Lin
+// Copyright 2014 Cisco Systems Inc. All rights reserved.
+// Copyright 2019 Red Hat Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ovsdb
+
+import (
+	"errors"
+	"fmt"
+	"github.com/socketplane/libovsdb"
+	"sync"
+)
+
+// OVS driver state
+type OvsBridgeDriver struct {
+	// OVS client
+	ovsClient *libovsdb.OvsdbClient
+
+	// Name of the OVS bridge
+	OvsBridgeName string
+
+	// read/write lock for accessing the cache
+	lock sync.RWMutex
+}
+
+// Create a new OVS driver with Unix socket
+func NewOvsBridgeDriver(bridgeName string) (*OvsBridgeDriver, error) {
+	ovsDriver := new(OvsBridgeDriver)
+
+	ovsDB, err := libovsdb.ConnectWithUnixSocket("/var/run/openvswitch/db.sock")
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to ovsdb error: %v", err)
+	}
+
+	// Setup state
+	ovsDriver.ovsClient = ovsDB
+	ovsDriver.OvsBridgeName = bridgeName
+
+	bridgeExist, err := ovsDriver.IsBridgePresent(bridgeName)
+	if err != nil {
+		return nil, err
+	}
+
+	if !bridgeExist {
+		return nil, fmt.Errorf("failed to find bridge %s", bridgeName)
+	}
+
+	// Return the new OVS driver
+	return ovsDriver, nil
+}
+
+// Wrapper for ovsDB transaction
+func (self *OvsBridgeDriver) ovsdbTransact(ops []libovsdb.Operation) error {
+	// Perform OVSDB transaction
+	reply, _ := self.ovsClient.Transact("Open_vSwitch", ops...)
+
+	if len(reply) < len(ops) {
+		return errors.New("OVS transaction failed. Less replies than operations")
+	}
+
+	// Parse reply and look for errors
+	for _, o := range reply {
+		if o.Error != "" {
+			return errors.New("OVS Transaction failed err " + o.Error + "Details: " + o.Details)
+		}
+	}
+
+	// Return success
+	return nil
+}
+
+// **************** OVS driver API ********************
+// Create an internal port in OVS
+func (self *OvsBridgeDriver) CreatePort(intfName, contNetnsPath, contIfaceName, ovnPortName string, vlanTag uint) error {
+	opStr := "insert"
+
+	intfUuid, intfOp, err := createInterfaceOperation(intfName, ovnPortName)
+	if err != nil {
+		return err
+	}
+
+	portUuid, portOp, err := createPortOperation(intfName, contNetnsPath, contIfaceName, vlanTag, intfUuid)
+	if err != nil {
+		return err
+	}
+
+	// mutate the Ports column of the row in the Bridge table
```

Comment:
the previous two helpers are nice! could you change this last bit to attachPortOperation?